// @jimdo/ui/dist/es/components/Feedback/Modal/Modal.js
import { __assign } from '../../../_virtual/_tslib.js';
import { IconButton } from '../../Buttons/IconButton/IconButton.js';
import { iconClose } from '../../DataDisplay/Icon/icons.js';
import { Overlay } from '../../Utils/Overlay/Overlay.js';
import classNames from 'classnames';
import React from 'react';
import styles from './Modal.module.scss.js';

var Modal = function (props) {
  var _a;

  var modalClasses = classNames(styles.modal, (_a = {}, _a[styles.modalSize] = props.modalSize === 'large', _a[styles.modalBackground] = props.transparent === true, _a[styles.smoothScrolling] = props.scrollable === true, _a), props.className);
  var closeClasses = classNames(styles.modalClose, props.closeClassName);

  var handleCloseClick = function () {
    if (props.closeable && props.onCloseClick) {
      props.onCloseClick();
    }
  };

  return React.createElement(Overlay, __assign({}, props, {
    className: modalClasses,
    onRequestClose: handleCloseClick,
    isOpen: true,
    overlayClassName: styles.backdrop
  }), React.createElement(React.Fragment, null, props.children, props.closeable && React.createElement(IconButton, {
    icon: iconClose,
    onClick: props.onCloseClick,
    className: closeClasses,
    buttonStyle: "primary",
    "aria-label": "close modal",
    "aria-hidden": true
  })));
};

export { Modal };
