// @jimdo/ui/dist/es/components/Buttons/SocialButtons/buttonSocialConnect/ButtonSocial.js
import { Icon } from '../../../DataDisplay/Icon/Icon.js';
import { iconCheckmarkCircle, iconMail, iconAppleSign, iconGoogleSign, iconFacebookSign } from '../../../DataDisplay/Icon/icons.js';
import classNames from 'classnames';
import React from 'react';
import styles from './ButtonSocial.module.scss.js';

var ButtonSocial = function (props) {
  var _a, _b;

  var classes = classNames(styles.buttonSocialConnect, (_a = {}, _a[styles.buttonSocialConnectFacebook] = props.social === 'fb', _a[styles.buttonSocialConnectGoogle] = props.social === 'google', _a[styles.buttonSocialConnectApple] = props.social === 'apple', _a[styles.buttonSocialConnectEmail] = props.social === 'email', _a[styles.buttonSocialConnectConnected] = props.connected, _a[styles.noGutter] = props.noGutter, _a), props.className);
  var iconClasses = classNames(styles.buttonSocialConnectIcon, (_b = {}, _b[styles.buttonSocialConnectIconFacebook] = props.social === 'fb', _b[styles.buttonSocialConnectIconGoogle] = props.social === 'google', _b[styles.buttonSocialConnectIconApple] = props.social === 'apple', _b[styles.buttonSocialConnectIconEmail] = props.social === 'email', _b));

  var getIcon = function () {
    if (props.connected) {
      return React.createElement(Icon, {
        icon: iconCheckmarkCircle,
        className: iconClasses
      });
    }

    switch (props.social) {
      case 'fb':
        return React.createElement(Icon, {
          icon: iconFacebookSign,
          className: iconClasses
        });

      case 'google':
        return React.createElement(Icon, {
          icon: iconGoogleSign,
          className: iconClasses
        });

      case 'apple':
        return React.createElement(Icon, {
          icon: iconAppleSign,
          className: iconClasses
        });

      case 'email':
        return React.createElement(Icon, {
          icon: iconMail,
          className: iconClasses
        });
    }
  };

  var disabled = props.disabled || Boolean(props.connected);

  if (props.href) {
    return (// eslint-disable-next-line jsx-a11y/anchor-is-valid
      React.createElement("a", {
        type: "button",
        className: classes,
        href: !disabled ? props.href : undefined,
        onClick: props.onClick
      }, getIcon(), React.createElement("span", {
        className: styles.buttonSocialConnectText
      }, props.children))
    );
  }

  return React.createElement("button", {
    type: "button",
    className: classes,
    disabled: disabled,
    onClick: props.onClick
  }, getIcon(), React.createElement("span", {
    className: styles.buttonSocialConnectText
  }, props.children));
};

ButtonSocial.displayName = 'ButtonSocial';
export { ButtonSocial };
